apache2/msc_multipart.c

```diff
@@ -841,121 +841,129 @@ int multipart_init(modsec_rec *msr, char **error_msg) {
-                    return -1;
-                }
-            }
-        }
-
-        /* Have we seen the semicolon in the header? */
-        if (seen_semicolon == 0) {
-            msr->mpd->flag_missing_semicolon = 1;
-        }
+    if (msr->mpd->boundary == NULL) { /* Could not find boundary in the C-T header. */
+        msr->mpd->flag_error = 1;
 
-        b = strchr(msr->mpd->boundary + 8, '=');
-        if (b == NULL) {
-            msr->mpd->flag_error = 1;
-            *error_msg = apr_psprintf(msr->mp, "Multipart: Invalid boundary in C-T (malformed).");
+        /* Test for case-insensitive boundary. Allowed by the RFC but highly unusual. */
+        if (multipart_count_boundary_params(msr->mp, msr->request_content_type) > 0) {
+            *error_msg = apr_psprintf(msr->mp, "Multipart: Invalid boundary in C-T (case sensitivity).");
             return -1;
         }
 
-        /* Check parameter name ends well. */
-        if (b != (msr->mpd->boundary + 8)) {
-            /* Check all characters between the end of the boundary
-             * and the = character.
-             */
-            for (p = msr->mpd->boundary + 8; p < b; p++) {
-                if (isspace(*p)) {
-                    /* Flag for whitespace after parameter name. */
-                    msr->mpd->flag_boundary_whitespace = 1;
-                } else {
-                    msr->mpd->flag_error = 1;
-                    *error_msg = apr_psprintf(msr->mp, "Multipart: Invalid boundary in C-T (parameter name).");
-                    return -1;
-                }
-            }
+        *error_msg = apr_psprintf(msr->mp, "Multipart: Boundary not found in C-T.");
+        return -1;
+    }
+
+    /* Check for extra characters before the boundary. */
+    int seen_semicolon = 0;
+    for (char *p = (char *)(msr->request_content_type + 19); p < msr->mpd->boundary; p++) {
+        if (*p == ';') {
+            seen_semicolon = 1;
         }
+    }
 
-        b++; /* Go over the = character. */
-        len = strlen(b);
+    /* Have we seen the semicolon in the header? */
+    if (seen_semicolon == 0) {
+        msr->mpd->flag_missing_semicolon = 1;
+    }
 
-        /* Flag for whitespace before parameter value. */
-        if (isspace(*b)) {
-            msr->mpd->flag_boundary_whitespace = 1;
+    /* Find the end of boundary - more attributes such as 'charset' may follow it */
+    char *boundary_end = NULL;
+    int trailing_whitespaces = 0;
+    for (boundary_end = msr->mpd->boundary; *boundary_end != '\0'; ++boundary_end) {
+        if (*boundary_end == ';') {
+            break;
+        } else if (isspace(*boundary_end)) {
+            ++trailing_whitespaces;
+        } else {
+            trailing_whitespaces = 0;
         }
+    }
+    int boundary_size = (boundary_end - msr->mpd->boundary) - trailing_whitespaces;
+    if (boundary_size <= 9) { /* The length is less than that of the "boundary" name */
```

Comment:
I don't think this can happen, because we already did `msr->mpd->boundary = strstr(msr->request_content_type, "boundary");`, right?

Reply:
True; having looked at the code again, I see that this check is redundant. The code immediately following it will fail if the string only contains "boundary" and nothing else. Let me remove it. 

Thanks!